src/cascade/serialize/hdf.py

```diff
@@ -0,0 +1,286 @@
+    return read_group(hdf_group, read_var, groups)
+
+
+def read_grid_group(hdf_group, groups=None):
+    return read_group(hdf_group, read_smooth_grid, groups)
+
+
+def read_group(hdf_group, reader, groups=None):
+    """
+    Reads a DismodGroup of Var.
+
+    Args:
+        hdf_group (h5py.Group): The HDF Group into which to write.
+        reader (function): Reads whatever it is from the file.
+        groups (DismodGroups): A pre-existing DismodGroups.
+
+    Returns:
+        DismodGroups
+    """
+    dismod_group = groups if groups else DismodGroups()
+    for group_name, group in dismod_group.items():
+        datasets = [ds_name for ds_name in hdf_group.keys() if ds_name.startswith(group_name)]
+        for ds_name in datasets:
+            ds = hdf_group[ds_name]
+            if "dismod_group" in ds.attrs:
+                key0 = ds.attrs["key0"]
+                if "key1" in ds.attrs:
+                    try:
+                        key = (str(key0), int(ds.attrs["key1"]))
+                    except ValueError:
+                        key = (str(key0), str(ds.attrs["key1"]))
+                elif group_name == "random_effect":
+                    key = (str(key0), None)
+                else:  # Rate
+                    key = str(key0)
+                group[key] = reader(ds)
+    return dismod_group
+
+
+def write_dimension(hdf_group, dim_type, values):
+    """Given an HDF group, write a set of ages or times,
+    possibly sharing a set of ages or times that already exist."""
+    # h5py recommends using S for arrays of fixed-size strings.
+    if isinstance(values, np.ndarray) and not np.issubdtype(values.dtype, np.number):
+        values = values.astype("S")
+    else:
+        if len(values) > 0 and isinstance(values[0], str):
+            values = np.array(values, dtype="S")
+        else:
+            values = np.array(values, dtype=np.float)
+
+    base_name = f"{dim_type}{len(values)}"
+    dim_idx = 0
+    dim_name = f"{base_name}_{dim_idx}"
+    while dim_name in hdf_group:
+        compare_axis = hdf_group[dim_name]
+        if np.issubdtype(values.dtype, np.number):
+            if np.allclose(values, compare_axis):
+                return compare_axis
+        elif np.all(values == compare_axis):
+            return compare_axis
+        dim_idx += 1
+        dim_name = f"{base_name}_{dim_idx}"
+
+    dim_value = hdf_group.create_dataset(dim_name, data=values)
+    return dim_value
+
+
+def write_var(hdf_group, var, name):
+    """Writes a var into an HDF group by writing an age, time, and
+    values datasets."""
+    ages = var.ages
+    times = var.times
+
+    ds = hdf_group.create_dataset(name, (len(ages), len(times)), dtype=np.float)
+    for aidx, tidx in product(range(len(ages)), range(len(times))):
+        ds[aidx, tidx] = var[ages[aidx], times[tidx]]
+
+    for scale_idx, kind in enumerate(["age", "time"]):
+        dimension = write_dimension(hdf_group, kind, [ages, times][scale_idx])
```

Comment:
`[ages, times][scale_idx]` This is super weird. I actually think I'd prefer `enumerate(zip([ages, times], ["age", "time"]))` ore something. Though I think you can make a pretty strong argument that that is also super weird.

Reply:
Went with a namedtuple with two members: name and coord. Looking better, but it does push it towards index 0 and index 1 instead of age and time.